### DISCUSSION TITLE
what is the best way to build a TUI for an existing REPL

### DISCUSSION THREAD

**Discussion by TomMelt** [Textualize/textual#4326] [General]
I have an open-source mpi debugger that I am working on called [mdb](https://github.com/TomMelt/mdb/tree/add-client-server).

It has a REPL-like input that is based on pythons [cmd](https://docs.python.org/3/library/cmd.html) module.

I use `asyncio` and my own client server to send commands from the users input, to the debug processes. I then print the debug output to the screen.

I want to create a TUI interface for my code, mostly because I want to view source code as I step through the debugger.

My initial thought was to use an [Input](https://textual.textualize.io/widgets/input/) widget. But then I lose my REPL's functionality (e.g., most importantly my readline shortcuts). Is there a widget that I can _run_ a terminal/REPL in.

It would be great if I could just call my [shell cmd loop](https://github.com/TomMelt/mdb/blob/de35bb88dc474dcfb8412d45712816f848b8f7de/mdb/mdb_attach.py#L143) from the TUI and forward the input to the REPL.

I have a layout for the TUI.
![image](https://github.com/Textualize/textual/assets/25719944/d41a16c2-50d8-40bb-92a4-adfb2b6ecc9b)

My plan is:
* REPL at the bottom
* file source at top
* command output in middle (or it could be in the same thing as the input if that's easier

**Comment by TomMelt** [comment]
I have rephrased the title of this discussion. 

I'm thinking that patching the `cmd` module in python might be an option too.

Would love to hear if anyone else has better suggestions

**Comment by brandon-fryslie** [comment]
Nothing?  This has to be a pretty common use case
